src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+    I = typeof(indices)
+    _RollingCutOut{A, I, T}(array, indices)
+end
+
+Base.size(r::_RollingCutOut) = (length(r.eachindex),)
+
+function Base.eachindex(r::_RollingCutOut, rs::_RollingCutOut...)
+    for r2 in rs
+        @assert r.eachindex == r2.eachindex
+    end
+    Base.IdentityUnitRange(r.eachindex)
+end
+Base.axes(r::_RollingCutOut) = (eachindex(r),)
+
+@inline function Base.getindex(o::_RollingCutOut, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i]
+end
+
+@inline function Base.setindex!(o::_RollingCutOut, val, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i] = val
+end
+
+############################## tmap, tmap! ##############################
+struct MapWorkspace{F,B,AD,AS}
+    f::F
+    batches::B
+    arena_dst_view::AD
+    arena_src_views::AS
+end
+
+@noinline function run!(o::MapWorkspace)
+    let o=o
+        Threads.@threads for i in 1:length(o.batches)
+            tid = Threads.threadid()
+            dst_view  = o.arena_dst_view[tid]
+            src_views = o.arena_src_views[tid]
+            inds = o.batches[i]
+            dst_view.eachindex = inds
+            for view in src_views
+                view.eachindex = inds
+            end
+            Base.map!(o.f, dst_view, src_views...)
+        end
+    end
+end
+
+function create_arena_src_views(srcs, sample_inds)
+    nt = Threads.nthreads()
+    [Base.map(src -> _RollingCutOut(src, sample_inds), srcs) for _ in 1:nt]
+end
+
+@noinline function prepare(::typeof(tmap!), f, dst, srcs; batch_size::Int)
+    # we use IndexLinear since _RollingCutOut implementation
+    # does not support other indexing well
+    all_inds  = eachindex(IndexLinear(), srcs...)
+    batches   = Batches(all_inds, batch_size)
+    sample_inds = batches[1]
+    nt = Threads.nthreads()
+    arena_dst_view  = [_RollingCutOut(dst, sample_inds) for _ in 1:nt]
+    arena_src_views = create_arena_src_views(srcs, sample_inds)
+    return MapWorkspace(f, batches, arena_dst_view, arena_src_views)
+end
+
+@noinline function tmap!(f, dst, srcs::AbstractArray...;
+                        batch_size=default_batch_size(length(dst)))
+    isempty(first(srcs)) && return dst
+    w = prepare(tmap!, f, dst, srcs, batch_size=batch_size)
+    run!(w)
+    dst
+end
+
+function tmap(f, srcs::AbstractArray...; 
+             batch_size=default_batch_size(length(first(srcs)))
+            )
+    g = Base.Generator(f,srcs...)
+    T = Base.@default_eltype(g)
+    dst = similar(first(srcs), T)
+    tmap!(f, dst, srcs...; batch_size=batch_size)
+end
+
+############################## tmapreduce(like) ##############################
+struct Reduction{O}
+    op::O
+end
+(red::Reduction)(f, srcs...) = Base.mapreduce(f, red.op, srcs...)
+
+struct MapReduceWorkspace{R,F,B,V,OA<:OffsetArray}
+    reduction::R
+    f::F
+    batches::Batches{B}
+    arena_src_views::V
+    batch_reductions::OA
+end
+
+struct NoInit end
+
+function create_reduction(::typeof(tmapreduce), op)
```

Comment:
Only one method, one line, and one use, so may be just inline the body where it is used?

Reply:
Good catch this is overengineered.